host/class/uvc/usb_host_uvc_2/include/usb/uvc_host.h

```diff
@@ -0,0 +1,236 @@
+                                          To address this, either optimize your processing speed or increase the `number_of_frame_buffers` parameter in
+                                          `uvc_host_stream_config_t.advanced` to allocate additional buffers. */
+};
+
+/**
+ * @brief UVC Device Event data structure
+ */
+typedef struct {
+    enum uvc_host_dev_event type;
+    union {
+        esp_err_t error;                   //!< Error code from USB Host
+        uvc_host_stream_hdl_t stream_hdl;  //!< Disconnection event
+    } data;
+} uvc_host_stream_event_data_t;
+
+/**
+ * @brief Formats supported by this driver
+ */
+enum uvc_host_stream_format {
+    UVC_VS_FORMAT_UNDEFINED = 0, // Invalid format. Do not request this format from the camera.
+    UVC_VS_FORMAT_MJPEG,
+    UVC_VS_FORMAT_YUY2,
+    UVC_VS_FORMAT_H264,
+    UVC_VS_FORMAT_H265,
+};
+
+typedef struct {
+    unsigned h_res;                     /**< Horizontal resolution */
+    unsigned v_res;                     /**< Vertical resolution */
+    float fps;                          /**< Frames per second */
+    enum uvc_host_stream_format format; /**< Frame coding format */
+} uvc_host_stream_format_t;
+
+/**
+ * @brief Video Stream frame
+ *
+ * This type is returned from frame callback upon receiving new frame
+ */
+typedef struct {
+    const uvc_host_stream_format_t vs_format; /**< Format of this frame buffer */
+    size_t data_buffer_len;                   /**< Max data length supported by this frame buffer */
+    size_t data_len;                          /**< Data length of currently store frame */
+    uint8_t *data;                            /**< Frame data */
+} uvc_host_frame_t;
+
+/**
+ * @brief Stream event callback type
+ *
+ * @param[in] event    Event structure
+ * @param[in] user_ctx User's argument passed to open function
+ */
+typedef void (*uvc_host_stream_callback_t)(const uvc_host_stream_event_data_t *event, void *user_ctx);
+
+/**
+ * @brief Frame callback type
+ *
+ * @param[in] frame    Received frame
+ * @param[in] user_ctx User's argument passed to open function
+ * @return
+ *     - true:  The frame was processed by user. The UVC driver owns the frame now.
+ *     - false: The frame was not yet processed by user.
+ *              The user must call uvc_host_frame_return() to return it to UVC driver.
+ */
+typedef bool (*uvc_host_frame_callback_t)(const uvc_host_frame_t *frame, void *user_ctx);
+
+/**
+ * @brief Configuration structure of UVC device
+ */
+typedef struct {
+    uvc_host_stream_callback_t event_cb;  /**< Stream's event callback function. Can be NULL */
+    uvc_host_frame_callback_t frame_cb;   /**< Stream's frame callback function */
+    void *user_ctx;                       /**< User's argument that will be passed to the callbacks */
+    struct {
+        uint16_t vid;                     /**< Device's Vendor ID. Set to 0 for any */
```

Comment:
If we are providing the option to select any vid/pid, it seems that it is a command  like: "I don't know anything about device will be attached, I just want to have a stream".
In this case, maybe it will be a good thing to add the default config as well.

Like literally "open the first device, first stream with the first possible supported settings and give me the handle". 

Because I've tried to provide correct config to open the stream and I realized that I have no idea about the VID/PID, stream number and protocols/formats the camera supports. 
So either I need to attach camera somewhere in advance to get these parameters, or just ask the driver to use the first supported configuration it'll find.

Reply:
> I don't know anything about device will be attached, I just want to have a stream

This is a very common usecase. That is why we allow VID/PID to be 0.

I checked how we could easily define some defaults for video stream and it would require more changes. The whole 'format negotiation' has very light implementation in current version of the driver. 

We can enhance it, as described in the TODO https://github.com/espressif/esp-usb/blob/104b5c141257423180efecb671197e12ac8518e8/host/class/uvc/usb_host_uvc_2/TODO#L34

Reply:
Cool, IMHO that would be very helpful!